File: sable-shared/src/errors/AppError.ts
// Base error class for all Sable services.
//
// Every operational error (auth failure, validation, downstream timeout, …)
// throws an AppError with a semantic code. HTTP status is looked up from the
// codes table — same code always maps to the same status. The error
// middleware in each service formats `error.toJSON()` and sends it back.

import { ERROR_CODES, type ErrorCode } from './codes.js';

export interface AppErrorOptions {
  /** Override the default message for this code. */
  message?: string;
  /** Structured context safe to send to the client (no PII, no internals). */
  details?: Record<string, unknown>;
  /** Original error that triggered this — preserved on `error.cause`. */
  cause?: unknown;
  /** Override the default HTTP status for this code. Rarely needed. */
  statusCode?: number;
}

export class AppError extends Error {
  readonly code: ErrorCode;
  readonly statusCode: number;
  readonly details?: Record<string, unknown>;

  constructor(code: ErrorCode, options: AppErrorOptions = {}) {
    const def = ERROR_CODES[code];
    const message = options.message ?? def.message;
    super(message, options.cause !== undefined ? { cause: options.cause } : undefined);

    this.name = 'AppError';
    this.code = code;
    this.statusCode = options.statusCode ?? def.status;
    if (options.details !== undefined) this.details = options.details;

    // Trim our constructor frames out of the stack on V8.
    if (typeof Error.captureStackTrace === 'function') {
      Error.captureStackTrace(this, this.constructor);
    }
  }

  /**
   * JSON-safe representation of the error body. Excludes `stack` and `cause` —
   * those are for logs, never for clients. Composed by `failure()` in
   * http/response.ts into the envelope.
   */
  toJSON(): { code: ErrorCode; message: string; details?: Record<string, unknown> } {
    return {
      code: this.code,
      message: this.message,
      ...(this.details !== undefined ? { details: this.details } : {}),
    };
  }

  /** Convenience guard so callers can `if (AppError.is(e))` without instanceof imports. */
  static is(value: unknown): value is AppError {
    return value instanceof AppError;
  }
}

File: sable-shared-py/sable_shared/errors/app_error.py
"""Operational error with an HTTP status and a wire-safe code.

Constructed from an ErrorCode (the table in ``codes.py`` supplies the
default status and message). Callers can override message and attach
``details`` for client-safe context — ``cause`` is preserved on
``__cause__`` for logging but never serialised to the wire.

Mirrors ``AppError`` in TS ``sable-shared/src/errors/AppError.ts``.
"""

from __future__ import annotations

from typing import Any

from sable_shared.errors.codes import ERROR_CODES, ErrorCode


class AppError(Exception):
    code: ErrorCode
    status_code: int
    message: str
    details: dict[str, Any] | None

    def __init__(
        self,
        code: ErrorCode,
        *,
        status_code: int | None = None,
        message: str | None = None,
        details: dict[str, Any] | None = None,
        cause: BaseException | None = None,
    ) -> None:
        entry = ERROR_CODES.get(code)
        if entry is None:
            # Unknown code at runtime — preserve it but fall back to 500.
            self.code = code
            self.status_code = status_code if status_code is not None else 500
            self.message = message or code.replace("_", " ").lower()
        else:
            self.code = code
            self.status_code = status_code if status_code is not None else entry["status"]
            self.message = message if message is not None else entry["message"]

        self.details = details
        if cause is not None:
            self.__cause__ = cause
        super().__init__(self.message)

    def to_dict(self) -> dict[str, Any]:
        """Wire-safe representation. Excludes cause / traceback."""
        body: dict[str, Any] = {"code": self.code, "message": self.message}
        if self.details is not None:
            body["details"] = self.details
        return body

    @staticmethod
    def is_app_error(value: object) -> bool:
        return isinstance(value, AppError)
